### PR TITLE
Introduce tooling api module

### PR DESCRIPTION
**Pull request by arturbosch**
Sorry for the huge PR upfront. Please review this commit by commit.

I want to make this right without a breaking change and an API which will work for the gradle, sonar, maven, bazel, intellij and cli integrations of the core module.

Why an extra tooling api?
- All mentioned integrations use `core` directly via the horrible `ProcessingSettings` class.
- `cli` is/(was) more `core` than `core`
- #2860

Notable additional changes in this PR:
- Running a specified single rule is supported by the tooling api. Cli's SingleRunner implementation was removed.
- KtFileModifier is now just an extension and has no special treatment in the core classes.
- `ProcessingSettings` is still `public` but only used internally in `core`. We can better refactor/remove this class in the future. Edit: I've refactored the class.
- `core` decides when to fail the build via a `MaxIssuePolicy` sealed class.
- `UnstableApi` annotation now has a `reason` property.
- `createRunner` test helper which uses `NullPrintStream` by default.
- `formatting` module does not depend on `core` anymore!

Usages:
- cli: in this PR
- gradle-plugin: TODO
- intellij-plugin: https://github.com/detekt/detekt-intellij-plugin/pull/94
- sonar-detekt: https://github.com/detekt/sonar-kotlin/pull/125
- compiler-plugin: https://github.com/detekt/detekt-compiler-plugin/pull/1

### CODE REVIEW

File: detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/runners/ConfigExporter.kt [hunk 1]
```diff
@@ -1,16 +1,17 @@
...
+    private val arguments: CliArgs,
+    private val outputPrinter: Appendable
+) : Executable {
 
     override fun execute() {
-        val configPath = arguments.config ?: DEFAULT_CONFIG
-        val defaultConfig = ClasspathResourceConverter().convert(DEFAULT_CONFIG).openStream()
-        defaultConfig.copyTo(File(configPath).outputStream())
-        println("Successfully copied default config to ${File(configPath).absolutePath}")
+        val configPath = Paths.get(arguments.config ?: "detekt.yaml")
```
Comment (BraisGabin) [inline detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/runners/ConfigExporter.kt:13]: `yaml` or `yml`? I would use `yml` to be consistent with the other files.

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/KtTreeCompiler.kt [hunk 2]
```diff
@@ -1,25 +1,32 @@
...
 
+    private val basePath: Path? = spec.basePath
+    private val pathFilters: PathFilters? =
+        PathFilters.of(spec.includes.toList(), spec.excludes.toList())
+
     companion object {
         val KT_ENDINGS = setOf("kt", "kts")
-        fun instance(settings: ProcessingSettings): KtTreeCompiler = KtTreeCompiler(settings)
+        fun instance(settings: ProcessingSettings) = KtTreeCompiler(settings, settings.spec.projectSpec)
     }
```
Comment (BraisGabin) [inline detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/KtTreeCompiler.kt:24]:
```kotlin
class KtTreeCompiler(
    private val settings: ProcessingSettings,
    private val compiler: KtCompiler = KtCompiler(settings.environment)
) {

    private val basePath: Path? = settings.spec.projectSpec.basePath
    private val pathFilters: PathFilters? =
        PathFilters.of(settings.spec.projectSpec.includes.toList(), settings.spec.projectSpec.excludes.toList())

    companion object {
        val KT_ENDINGS = setOf("kt", "kts")
    }
```

We don't need any factory here, right?

File: detekt-cli/build.gradle.kts [hunk 3]
```diff
@@ -5,6 +5,7 @@ application {
 dependencies {
     implementation(kotlin("stdlib-jdk8"))
     implementation(project(":detekt-core"))
```
[the remark this answers is not in the document]
Comment (arturbosch) [reply detekt-cli/build.gradle.kts]: Yes, thanks, done :).

File: detekt-core/build.gradle.kts [hunk 4]
```diff
@@ -3,6 +3,7 @@ dependencies {
     api(project(":detekt-metrics"))
     api(project(":detekt-parser"))
     api(project(":detekt-psi-utils"))
+    api(project(":detekt-tooling"))
```
[the remark this answers is not in the document]
Comment (arturbosch) [reply detekt-core/build.gradle.kts:6]: Everything can be `implementation` now.

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/config/DefaultConfig.kt [hunk 5]
```diff
@@ -8,7 +8,7 @@ object DefaultConfig {
     const val RESOURCE_NAME = "default-detekt-config.yml"
```
[the remark this answers is not in the document]
Comment (arturbosch) [reply detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/config/DefaultConfig.kt:8]: It was used by the cli. However I made `core` a runtimeOnly dependency and hardcoded the same string.
It's okay as we should remove the cli `NotApiButProbablyUsed` after 2 more minor versions.

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/tooling/DefaultAnalysisResult.kt [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+@OptIn(UnstableApi::class)
+class DefaultAnalysisResult(
+    override val container: Detektion?,
+    override val error: DetektError? = null
+) : AnalysisResult {
+
+    init {
+        require(container != null || error != null)
+    }
+}
```
[the remark this answers is not in the document]
Comment (arturbosch) [reply detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/tooling/DefaultAnalysisResult.kt]: Not sure, could be good or overengineered :). There are three cases for `AnalysisResult`:

```kotlin
@UnstableApi
sealed class AnalysisResultNew(
    override val error: DetektError?,
    override val container: Detektion?
) : AnalysisResult {

    class Failure(override val error: DetektError) : AnalysisResultNew(error, null)
    class Success(override val container: Detektion) : AnalysisResultNew(null, container)
    class ResultWithError(
        override val container: Detektion,
        override val error: DetektError
    ) : AnalysisResultNew(error, container)
}
```

What do you think?

File: detekt-tooling/src/main/kotlin/io/github/detekt/tooling/api/Detekt.kt [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+import org.jetbrains.kotlin.psi.KtFile
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.BindingTrace
+import java.nio.file.Path
+
+/**
+ * Instance of detekt.
+ *
+ * Runs analysis based on [io.github.detekt.tooling.api.spec.ProcessingSpec] configuration.
```
[the remark this answers is not in the document]
Comment (arturbosch) [reply detekt-tooling/src/main/kotlin/io/github/detekt/tooling/api/Detekt.kt:11]: Well that class is part of the dsl.
`DetektProvider#get(ProcessingSpec)`.

**Comment by arturbosch** [comment]
I'm not sure why jacoco fails on detekt-core with Java8 but I can reproduce this locally...

**Comment by arturbosch** [comment]
> I'm not sure why jacoco fails on detekt-core with Java8 but I can reproduce this locally...

This is a really strange one!
The stacktrace on the console is:

2020-07-12-214325_1302x228_scrot

But tacking a look at the gradle test report I get:

```
org.gradle.api.internal.tasks.testing.TestSuiteExecutionException: Could not complete execution for Gradle Test Executor 15.
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:63)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.stop(TestWorker.java:133)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NoClassDefFoundError: org/junit/platform/commons/util/BlacklistedExceptions
	at org.junit.platform.launcher.core.DefaultLauncher.discoverEngineRoot(DefaultLauncher.java:187)
	at org.junit.platform.launcher.core.DefaultLauncher.discoverRoot(DefaultLauncher.java:168)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:132)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	... 25 more
Caused by: java.lang.ClassNotFoundException: org.junit.platform.commons.util.BlacklistedExceptions
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	... 32 more
```

So Gradle can't find `org.junit.platform.commons.util.BlacklistedExceptions` at runtime.

Now I stumbled upon this issue https://github.com/junit-team/junit5/pull/2322/files ....

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/tooling/DefaultAnalysisResult.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (BraisGabin) [reply detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/tooling/DefaultAnalysisResult.kt]: Oh, if there are three options maybe it's better as it's now. I was thinking that this was some kind of `Either`.

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/reporting/Colorizer.kt [hunk 8]
```diff
@@ -14,7 +14,7 @@ private data class Color(private val value: Byte) {
         get() = "$ESC[${value}m"
 }
 
-private val isColoredOutputSupported: Boolean = !IS_WINDOWS
+private val isColoredOutputSupported: Boolean = !SystemInfo.isWindows
```
Comment (BraisGabin) [inline detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/reporting/Colorizer.kt:17]: nice!

File: detekt-tooling/src/main/kotlin/io/github/detekt/tooling/api/Detekt.kt [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+ */
+interface Detekt {
+
+    fun run(): AnalysisResult
+
+    fun run(path: Path): AnalysisResult
+
+    fun run(sourceCode: String, filename: String): AnalysisResult
+
+    fun run(files: Collection<KtFile>, bindingContext: BindingContext): AnalysisResult
```
Comment (BraisGabin) [inline detekt-tooling/src/main/kotlin/io/github/detekt/tooling/api/Detekt.kt:20]: Could we reduce this API even more? Would it help?

Comment (arturbosch) [reply detekt-tooling/src/main/kotlin/io/github/detekt/tooling/api/Detekt.kt:20]: These are the minimal variants we need for our plugins.
IDE's/the IntelliJ plugin may want to test kotlin code which is not yet on hard-disc -> choose string or path variant
The compiler plugin api gives us the resolved BindingContext.
Just run will use the `inputPaths`.

**Comment by arturbosch** [comment]
I've fixed the build/classloader issue.
The PR is ready.
I will try to increase the code coverage now.

**Comment by BraisGabin** [comment]
Should we deprecate `Main.buildRunner()` from `:detekt-cli`?

**Comment by arturbosch** [comment]
> Should we deprecate `Main.buildRunner()` from `:detekt-cli`?

Already done :)

https://github.com/detekt/detekt/blob/73149280ef1d4d30693fffe2ec07647d5d7ecb7f/detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/Main.kt#L39-L47

**Comment by BraisGabin** [comment]
> > Should we deprecate `Main.buildRunner()` from `:detekt-cli`?
> 
> Already done :)
> 
> https://github.com/detekt/detekt/blob/73149280ef1d4d30693fffe2ec07647d5d7ecb7f/detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/Main.kt#L39-L47

ups! my bad. I had this as an annotation in my note book from when I did the code review and I forgot to comment it at that moment and then I didn't recheck.

**Pull request merged**